examples/async.rs

```diff
@@ -0,0 +1,294 @@
+        let mut events = EventStream::new();
+
+        while !self.should_quit {
+            tokio::select! {
+                _ = interval.tick() => self.draw(&mut terminal)?,
+                Some(Ok(event)) = events.next() =>  self.handle_event(&event),
+            }
+        }
+        Ok(())
+    }
+
+    fn draw(&self, terminal: &mut Terminal) -> Result<()> {
+        terminal.draw(|frame| {
+            let area = frame.size();
+            frame.render_widget(
+                Line::from("ratatui async example").centered().cyan().bold(),
+                area,
+            );
+            let area = area.offset(Offset { x: 0, y: 1 }).intersection(area);
+            frame.render_widget(&self.pulls, area);
+        })?;
+        Ok(())
+    }
+
+    fn handle_event(&mut self, event: &Event) {
+        if let Event::Key(event) = event {
+            match event.code {
+                KeyCode::Char('q') => self.should_quit = true,
+                KeyCode::Char('j') => self.pulls.scroll_down(),
+                KeyCode::Char('k') => self.pulls.scroll_up(),
+                _ => {}
+            }
+        }
+    }
+}
+
+/// A widget that displays a list of pull requests.
+///
+/// This is an async widget that fetches the list of pull requests from the GitHub API. It contains
+/// an inner `Arc<RwLock<PullRequests>>` that holds the state of the widget. Cloning the widget
+/// will clone the Arc, so you can pass it around to other threads, and this is used to spawn a
+/// background task to fetch the pull requests.
+#[derive(Debug, Clone, Default)]
+struct PullRequestsWidget {
+    inner: Arc<RwLock<PullRequests>>,
+    selected_index: usize, // no need to lock this since it's only accessed by the main thread
+}
+
+#[derive(Debug, Default)]
+struct PullRequests {
+    pulls: Vec<PullRequest>,
+    loading_state: LoadingState,
+}
+
+#[derive(Debug, Clone)]
+struct PullRequest {
+    id: String,
+    title: String,
+    url: String,
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+enum LoadingState {
+    #[default]
+    Idle,
+    Loading,
+    Loaded,
+    Error(String),
+}
+
+impl PullRequestsWidget {
+    /// Start fetching the pull requests in the background.
+    ///
+    /// This method spawns a background task that fetches the pull requests from the GitHub API.
+    /// The result of the fetch is then passed to the `on_load` or `on_err` methods.
+    fn run(&self) {
+        let this = self.clone(); // clone the widget to pass to the background task
+        tokio::spawn(async move {
+            // this runs once, but you could also run this in a loop, using a channel that accepts
+            // messages to refresh on demand, or with an interval timer to refresh every N seconds
+            this.set_loading_state(LoadingState::Loading);
+            match fetch_pulls().await {
+                Ok(page) => this.on_load(&page),
+                Err(err) => this.on_err(&err),
+            }
```

Comment:
As an example of why I like inlining `tokio::time::interval`, in an example like this, I'm looking for where are the futures that are being `.await`ed on, and in which parts of the code are they in. Here I like that this code has `tokio::spawn(...)`. 

If I were to go too far, I'd say I even prefer `fetch_pulls` to be inlined or atleast defined inside this `run` method.

Reply:
e658be372